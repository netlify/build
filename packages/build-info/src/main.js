import { getContext } from './context.js'
import { buildInfo } from './core.js'

export const getBuildInfo = async function (opts) {
  const context = await getContext(opts)
  return await buildInfo(context)
}
