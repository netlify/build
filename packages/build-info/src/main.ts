import { ContextOptions, getContext } from './context.js'
import { buildInfo } from './core.js'

export const getBuildInfo = async (opts: ContextOptions) => {
  const context = await getContext(opts)
  return await buildInfo(context)
}
