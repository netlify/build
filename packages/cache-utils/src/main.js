'use strict'

require('./utils/polyfills')

const del = require('del')

const { getCacheDir } = require('./dir')
const { moveCacheFile, hasFiles } = require('./fs')
const { list } = require('./list')
const { getManifestInfo, writeManifest, removeManifest, isExpired } = require('./manifest')
const { parsePath } = require('./path')

// Cache a file
const saveOne = async function (
  path,
  { move = DEFAULT_MOVE, ttl = DEFAULT_TTL, digests = [], cacheDir, cwd: cwdOpt, mode } = {},
) {
  const { srcPath, cachePath } = await parsePath({ path, cacheDir, cwdOpt, mode })

  if (!(await hasFiles(srcPath))) {
    return false
  }

  const { manifestInfo, identical } = await getManifestInfo({ cachePath, move, ttl, digests })
  if (identical) {
    return true
  }

  await del(cachePath, { force: true })
  await moveCacheFile(srcPath, cachePath, move)
  await writeManifest(manifestInfo)

  return true
}

// Restore a cached file
const restoreOne = async function (path, { move = DEFAULT_MOVE, cacheDir, cwd: cwdOpt, mode } = {}) {
  const { srcPath, cachePath } = await parsePath({ path, cacheDir, cwdOpt, mode })

  if (!(await hasFiles(cachePath))) {
    return false
  }

  if (await isExpired(cachePath)) {
    return false
  }

  await del(srcPath, { force: true })
  await moveCacheFile(cachePath, srcPath, move)

  return true
}

// Remove the cache of a file
const removeOne = async function (path, { cacheDir, cwd: cwdOpt, mode } = {}) {
  const { cachePath } = await parsePath({ path, cacheDir, cwdOpt, mode })

  if (!(await hasFiles(cachePath))) {
    return false
  }

  await del(cachePath, { force: true })
  await removeManifest(cachePath)

  return true
}

// Check if a file is cached
const hasOne = async function (path, { cacheDir, cwd: cwdOpt, mode } = {}) {
  const { cachePath } = await parsePath({ path, cacheDir, cwdOpt, mode })

  return (await hasFiles(cachePath)) && !(await isExpired(cachePath))
}

const DEFAULT_MOVE = false
const DEFAULT_TTL = undefined

// Allow each of the main functions to take either a single path or an array of
// paths as arguments
const allowMany = async function (func, paths, ...args) {
  if (!Array.isArray(paths)) {
    return func(paths, ...args)
  }

  const results = await Promise.all(paths.map((path) => func(path, ...args)))
  return results.some(Boolean)
}

const save = allowMany.bind(null, saveOne)
const restore = allowMany.bind(null, restoreOne)
const remove = allowMany.bind(null, removeOne)
const has = allowMany.bind(null, hasOne)

// Change `opts` default values
const bindOpts = function (opts) {
  return {
    save: (paths, optsA) => save(paths, { ...opts, ...optsA }),
    restore: (paths, optsA) => restore(paths, { ...opts, ...optsA }),
    remove: (paths, optsA) => remove(paths, { ...opts, ...optsA }),
    has: (paths, optsA) => has(paths, { ...opts, ...optsA }),
    list: (optsA) => list({ ...opts, ...optsA }),
    getCacheDir: (optsA) => getCacheDir({ ...opts, ...optsA }),
  }
}

module.exports = { save, restore, remove, has, list, getCacheDir, bindOpts }
