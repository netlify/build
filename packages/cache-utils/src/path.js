'use strict'

const { homedir } = require('os')
const { resolve, isAbsolute, join, sep } = require('path')

const { getCacheDir } = require('./dir')
const { safeGetCwd } = require('./utils/cwd')

// Find the paths of the file before/after caching
const parsePath = async function ({ path, cacheDir, cwdOpt, mode }) {
  const srcPath = await getSrcPath(path, cwdOpt)
  const cachePath = await getCachePath({ srcPath, cacheDir, cwdOpt, mode })
  return { srcPath, cachePath }
}

// Retrieve absolute path to the local file to cache/restore
const getSrcPath = async function (path, cwdOpt) {
  const cwd = await safeGetCwd(cwdOpt)
  const srcPath = resolvePath(path, cwd)
  checkSrcPath(srcPath, cwd)
  return srcPath
}

const resolvePath = function (path, cwd) {
  if (isAbsolute(path)) {
    return resolve(path)
  }

  if (cwd !== '') {
    return resolve(cwd, path)
  }

  throw new Error(`Current directory does not exist: ${cwd}`)
}

// Caching the whole repository creates many issues:
//  - It caches many directories that are not related to Gatsby but take lots of
//    space, such as node_modules
//  - It caches directories that are not meant to restored across builds. For
//    example .git (beside being big).
//  - It undoes any build operations inside the repository that might have
//    happened before this plugin starts restoring the cache, leading to
//    conflicts with other plugins, Netlify Build or the buildbot.
const checkSrcPath = function (srcPath, cwd) {
  if (cwd !== '' && isParentPath(srcPath, cwd)) {
    throw new Error(`Cannot cache ${srcPath} because it is the current directory (${cwd}) or a parent directory`)
  }
}

// Note: srcPath and cwd are already normalized and absolute
const isParentPath = function (srcPath, cwd) {
  return `${cwd}${sep}`.startsWith(`${srcPath}${sep}`)
}

const getCachePath = async function ({ srcPath, cacheDir, cwdOpt, mode }) {
  const cacheDirA = await getCacheDir({ cacheDir, mode })
  const { name, relPath } = await findBase(srcPath, cwdOpt)
  const cachePath = join(cacheDirA, name, relPath)
  return cachePath
}

// The cached path is the path relative to the base which can be either the
// current directory, the home directory or the root directory. Each is tried
// in order.
const findBase = async function (srcPath, cwdOpt) {
  const bases = await getBases(cwdOpt)
  const srcPathA = normalizeWindows(srcPath)
  return bases.map(({ name, base }) => parseBase(name, base, srcPathA)).find(Boolean)
}

// Windows drives are problematic:
//  - they cannot be used in `relPath` since directories cannot be called `C:`
//    for example.
//  - they make comparing between drives harder
// For the moment, we simply strip them. This does mean two files with the same
// paths but inside different drives would be cached together, which is not
// correct.
// This also means `cache.list()` always assumes the source files are on the
// current drive.
// TODO: fix it.
const normalizeWindows = function (srcPath) {
  return srcPath.replace(WINDOWS_DRIVE_REGEX, '\\')
}

const WINDOWS_DRIVE_REGEX = /^[a-zA-Z]:\\/

// This logic works when `base` and `path` are on different Windows drives
const parseBase = function (name, base, srcPath) {
  if (srcPath === base || !srcPath.startsWith(base)) {
    return
  }

  const relPath = srcPath.replace(base, '')
  return { name, relPath }
}

const getBases = async function (cwdOpt) {
  const cwdBase = await getCwdBase(cwdOpt)
  return [...cwdBase, { name: 'home', base: homedir() }, { name: 'root', base: sep }]
}

const getCwdBase = async function (cwdOpt) {
  const cwd = await safeGetCwd(cwdOpt)
  if (cwd === '') {
    return []
  }

  return [{ name: 'cwd', base: cwd }]
}

module.exports = { parsePath, getBases }
