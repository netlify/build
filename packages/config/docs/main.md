# Netlify configuration

## Purpose

The goal of `@netlify/config` is to abstract the logic that loads, normalizes and validates the Netlify configuration
(`netlify.toml`, UI build settings, `_headers`, `_redirects`, etc.). It returns a single normalized object to consumers.

Note: while `@netlify/build` share the same monorepo, it has a very distinct purpose.

## Consumers

At the moment, `@netlify/config` is used by four parts of Netlify's architecture. It might also be used by additional
consumers in the future, including outside of Netlify.

It is possible to distinguish between those different consumers using the [`mode` flag](../README.md#mode) which is set
to `"buildbot"`, `"cli"` or `"require"` by each of them. However, we try to limit the usage of that flag as much as
possible:

- Instead, we want those different modes to behave as similarly to each other as possible, especially Netlify CLI and
  buildbot.
- There are some cases where those different environments do require different logic. When this happens, we try to add
  new flags to toggle each specific behavior, as opposed to rely on the `mode` flag which would be less composable and
  re-usable.

### Programmatic vs CLI

There is both
[a CLI](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/bin/main.js#L16)
and
[programmatic entry point](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/main.js#L30).

To ensure both entry points re-use as much logic as possible, we try to minimize the amount of CLI-specific logic. At
the moment, the CLI code only includes some logic that makes sense exclusively within that context: exit code, CLI flags
parsing, terminal output.

By default, the CLI entry point prints to `stdout`. However,
[an `--output` flag](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/bin/main.js#L38)
is used
[by the buildbot](https://github.com/netlify/buildbot/blob/b6c6fc159f077f5dbda69ceb17363960232243fe/bot/configuration.go#L129)
to print to a file instead, which avoids crashing the process when the configuration object is very big.

### Buildbot

The buildbot logic uses multiple configuration properties: build directory, Functions directory, etc.Therefore,
[`@netlify/config` is called via the CLI](https://github.com/netlify/buildbot/blob/b6c6fc159f077f5dbda69ceb17363960232243fe/bot/configuration.go#L120)
(since the buildbot is written in a different language) at the beginning of each build.

`@netlify/config` is installed
[at Docker build time](https://github.com/netlify/buildbot/blob/b6c6fc159f077f5dbda69ceb17363960232243fe/package.json#L12)
by the buildbot. It is run with
[a hardcoded Node.js version](https://github.com/netlify/buildbot/blob/b6c6fc159f077f5dbda69ceb17363960232243fe/script/install-nvm.sh#L16).

### Netlify Build

`@netlify/build` also
[loads the configuration object](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/build/src/core/config.js#L107)
with `@netlify/config`, but programmatically.

Since the buildbot already loads that same object, we cache it
[as a `cachedConfig` object](https://github.com/netlify/buildbot/blob/b6c6fc159f077f5dbda69ceb17363960232243fe/script/run-build.sh#L75)
so `@netlify/build` does not need to re-compute it. This also ensures `@netlify/build` and the buildbot use the exact
same configuration properties.

#### Build plugins

Build plugins can both access and modify the Netlify configuration normalized object returned by `@netlify/config` using
[a `netlifyConfig` property passed to them](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/build/src/plugins/child/run.js#L16).

[When modified](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/build/src/plugins/child/diff.js#L21),
`@netlify/config`
[is called again](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/build/src/core/config.js#L126).
It is also called again when a build command or plugin
[adds a `_headers` or `_redirects` file](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/build/src/steps/update_config.js#L59).

The new configuration properties are printed, with some additional information when in debug mode.

The configuration mutations
[are stored in an `inlineConfig` object](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/mutations/apply.js#L14),
which is
[passed as a flag to `@netlify/config`](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/build/src/steps/update_config.js#L13)
and has the highest merging priority.

The logic in `@netlify/build` takes into account that the Netlify configuration can be mutated at any point. While most
configuration properties can be mutated,
[some are still read-only](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/mutations/apply.js#L52)
due to that additional complexity.

In production, any configuration mutation
[is temporarily persisted to `netlify.toml`](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/mutations/update.js#L17)
before deploying the site, and restored afterwards.

### Netlify CLI

Most Netlify CLI commands rely on configuration properties. The base command loads it
[using `@netlify/config` programmatically](https://github.com/netlify/cli/blob/f87d9e6e1749bedaabaace3dae98ca1ed6d84fd6/src/commands/base-command.js#L474).

An `inlineConfig` flag was added to `@netlify/config` so Netlify CLI can override specific configuration properties
based on its own CLI flags, but this is not used yet (see https://github.com/netlify/cli/issues/1265,
https://github.com/netlify/cli/issues/960, https://github.com/netlify/cli/issues/961).

We try to
[emulate the same environment variables](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/env/main.js#L13)
as on the buildbot, as much as possible.

### Automated tests

A few flags are available for programmatic users. Those are mostly used by our automated tests:

- [The `testOpts` flag](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/bin/flags.js#L153)
- [The `buffer` flag](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/bin/flags.js#L166)
  which allows returning the output programmatically instead of printing it on the terminal

## File paths

The Netlify configuration relies on many file paths. Those are documented [here](file_paths.md).

## UI build settings

Users can specify the configuration either in files (`netlify.toml`, `_headers`, `_redirects`) or in the Netlify UI
("build settings"). The latter has lower merging priority.

UI build settings are passed to `@netlify/config` using
[the `defaultConfig` flag](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/default.js#L6).

In the buildbot, this is populated
[with some values](https://github.com/netlify/buildbot/blob/7383bb8971c718de9618a700ec5049cb71af7102/bot/buildbot.go#L135-L144)
coming from the backend API.

When called from the Netlify CLI, API calls to the backend
[are being made](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/api/site_info.js#L24)
to retrieve
[those values](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/api/build_settings.js#L7).
This relies on multiple flags:

- The
  [`token`](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/api/client.js#L7)
  and
  [`siteId`](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/api/site_info.js#L30)
  flags require the site to have been linked locally and the user to be logged in.
- [The `host`, `scheme` and `pathPrefix` flags](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/api/client.js#L12)
  allow overriding the API endpoint URL.
- [The `offline` flag](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/api/client.js#L7)
  can disable this.

## Headers and redirects

Redirects configuration is more complicated:

- It has multiple properties
- It can be specified either in `netlify.toml` or in a separate `_redirects` file located in the publish directory and
  with its own Netlify-specific file format

Therefore, that logic is abstracted to
[its own module `netlify-redirect-parser`](https://github.com/netlify/netlify-redirect-parser).

Headers are handled very similarly to redirects, for consistency. They use a separate `_headers` file
[and `netlify-headers-parser` module](https://github.com/netlify/netlify-headers-parser).

## Context-specific configuration

Users can make some configuration specific to a given context or branch. Those are specified by embedding a
configuration object inside the main one, but namespaced with `context.{context|branch}.*`. `@netlify/config`
[merges those with higher priority](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/context.js#L28),
but only if the current context or branch matches.

Those are specified with:

- The `context` flag, which
  [defaults to `"production"`](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/options/main.js#L42)
- The `branch` flag,
  [which defaults](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/options/branch.js#L9)
  to the `BRANCH` environment variable, `HEAD` branch, `main` branch or `"master"` (this name was kept for backward
  compatibility).

## Normalization and validation

As described above, the final configuration object is a product of multiple merges:

- The configuration object loaded from:
  - Files (`config`)
  - The UI build settings (`defaultConfig`)
  - Build plugins modifications (`inlineConfig`)
- Context/branch-specific properties

As a result, normalization and validation happens
[in multiple stages](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/main.js#L259).
For example, we need to validate that `context` is a plain object before merging context properties. However, since the
build command might be specified either in `context` properties or not, we can only validate it after those have been
merged.

Additionally, we keep track of where the property was originally set
[with sibling properties named `origin`](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/origin.js#L11).
At the moment, this is mostly used for logging and debugging purposes. For example, the build logs show whether the
build command was set in `netlify.toml` or in the UI.

Almost all `@netlify/config` errors are user configuration errors. To help both users debug those problems and support
focus on actual system bugs, we enforce some strict configuration
[validation](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/validate/validations.js#L40).
Error messages are printed in the build logs with a nice format and examples on how to fix them. This logic relies on
declaring each configuration property's shape and example values
[in the following file](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/validate/validations.js#L40).
This requires some (sometimes tedious) maintenance, but improves the developer experience.

The configuration object is normalized, which includes:

- Assigning
  [default values](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/normalize.js#L22-L32)
- Allowing some properties
  [to be capitalized](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/case.js#L2),
  for backward compatibility
- [Normalizing the Functions-related properties](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/functions_config.js#L11),
  which are rather complex and polymorphic

## Debugging

A `debug` flag can be used
[to log the flags and configuration object](https://github.com/netlify/build/blob/558fe8869f47ecc84d05bd4d26d32df00f47a3b9/packages/config/src/log/main.js#L6)
at different stages of the logic. It can enabled with the `NETLIFY_BUILD_DEBUG` environment variable as well.

We make sure that no secret information is ever logged by this debugging feature.
