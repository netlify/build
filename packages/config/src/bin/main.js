#!/usr/bin/env node
const process = require('process')

const { stableStringify } = require('fast-safe-stringify')

require('../utils/polyfills')

const { isUserError } = require('../error')
const resolveConfig = require('../main')

const { parseFlags } = require('./flags')

// CLI entry point
const runCli = async function() {
  try {
    const { stable, ...flags } = parseFlags()
    const result = await resolveConfig(flags)
    handleCliSuccess(result, stable)
  } catch (error) {
    handleCliError(error)
  }
}

// The result is printed as JSON on stdout on success (exit code 0)
const handleCliSuccess = function(result, stable) {
  const resultA = serializeApi(result)
  const stringifyFunc = stable ? stableStringify : JSON.stringify
  const resultJson = stringifyFunc(resultA, null, 2)
  console.log(resultJson)
  process.exitCode = 0
}

// `api` is not JSON-serializable, so we remove it
// We still indicate it as a boolean
const serializeApi = function({ api, ...result }) {
  if (api === undefined) {
    return result
  }

  return { ...result, hasApi: true }
}

const handleCliError = function(error) {
  // Errors caused by users do not show stack traces and have exit code 1
  if (isUserError(error)) {
    console.error(error.message)
    process.exitCode = 1
    return
  }

  // Internal errors / bugs have exit code 2
  console.error(error.stack)
  process.exitCode = 2
}

runCli()
