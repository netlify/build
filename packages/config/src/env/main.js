import omit from 'omit.js'

import { removeFalsy } from '../utils/remove_falsy.js'

import { getGitEnv } from './git.js'

// Retrieve this site's environment variable. Also take into account team-wide
// environment variables and addons.
// The buildbot already has the right environment variables. This is mostly
// meant so that local builds can mimic production builds
// TODO: add `netlify.toml` `build.environment`, after normalization
// TODO: add `CONTEXT` and others
export const getEnv = async function ({
  mode,
  config,
  siteInfo,
  accounts,
  addons,
  buildDir,
  branch,
  deployId,
  buildId,
  context,
}) {
  if (mode === 'buildbot') {
    return {}
  }

  const generalEnv = await getGeneralEnv({ siteInfo, buildDir, branch, deployId, buildId, context })
  const [accountEnv, addonsEnv, uiEnv, configFileEnv] = getUserEnv({ config, siteInfo, accounts, addons })

  // Sources of environment variables, in descending order of precedence.
  const sources = [
    { key: 'configFile', values: configFileEnv },
    { key: 'ui', values: uiEnv },
    { key: 'addons', values: addonsEnv },
    { key: 'account', values: accountEnv },
    { key: 'general', values: generalEnv },
  ]

  // A hash mapping names of environment variables to objects containing the following properties:
  // - sources: List of sources where the environment variable was found. The first element is the source that
  //   actually provided the variable (i.e. the one with the highest precedence).
  // - value: The value of the environment variable.
  const env = new Map()

  sources.forEach((source) => {
    Object.keys(source.values).forEach((key) => {
      if (env.has(key)) {
        const { sources: envSources, value } = env.get(key)

        env.set(key, {
          sources: [...envSources, source.key],
          value,
        })
      } else {
        env.set(key, {
          sources: [source.key],
          value: source.values[key],
        })
      }
    })
  })

  return Object.fromEntries(env)
}

// Environment variables not set by users, but meant to mimic the production
// environment.
const getGeneralEnv = async function ({
  siteInfo: { id, name, url, build_settings: { repo_url: REPOSITORY_URL } = {} },
  buildDir,
  branch,
  deployId,
  buildId,
  context,
}) {
  const gitEnv = await getGitEnv(buildDir, branch)
  return removeFalsy({
    SITE_ID: id,
    SITE_NAME: name,
    DEPLOY_ID: deployId,
    BUILD_ID: buildId,
    URL: url,
    REPOSITORY_URL,
    CONTEXT: context,
    NETLIFY_LOCAL: 'true',
    ...gitEnv,
    // Localization
    LANG: 'en_US.UTF-8',
    LANGUAGE: 'en_US:en',
    LC_ALL: 'en_US.UTF-8',
    // Disable telemetry of some tools
    GATSBY_TELEMETRY_DISABLED: '1',
    NEXT_TELEMETRY_DISABLED: '1',
  })
}

// Environment variables specified by the user
const getUserEnv = function ({ config, siteInfo, accounts, addons }) {
  const accountEnv = getAccountEnv({ siteInfo, accounts })
  const addonsEnv = getAddonsEnv(addons)
  const uiEnv = getUiEnv({ siteInfo })
  const configFileEnv = getConfigFileEnv({ config })
  return [accountEnv, addonsEnv, uiEnv, configFileEnv].map(cleanUserEnv)
}

// Account-wide environment variables
const getAccountEnv = function ({ siteInfo: { account_slug: accountSlug }, accounts }) {
  const { site_env: siteEnv = {} } = accounts.find(({ slug }) => slug === accountSlug) || {}
  return siteEnv
}

// Environment variables from addons
const getAddonsEnv = function (addons) {
  return Object.assign({}, ...addons.map(getAddonEnv))
}

const getAddonEnv = function ({ env }) {
  return env
}

// Site-specific environment variables set in the UI
const getUiEnv = function ({ siteInfo: { build_settings: { env = {} } = {} } }) {
  return env
}

// Site-specific environment variables set in netlify.toml
const getConfigFileEnv = function ({
  config: {
    build: { environment = {} },
  },
}) {
  return environment
}

// Some environment variables cannot be overridden by configuration
const cleanUserEnv = function (userEnv) {
  return omit.default(userEnv, READONLY_ENV)
}

const READONLY_ENV = [
  // Set in local builds
  'BRANCH',
  'CACHED_COMMIT_REF',
  'COMMIT_REF',
  'CONTEXT',
  'HEAD',
  'REPOSITORY_URL',
  'URL',

  // CI builds set NETLIFY=true while CLI and programmatic builds set
  // NETLIFY_LOCAL=true
  'NETLIFY',
  'NETLIFY_LOCAL',

  // Not set in local builds because there is no CI build/deploy, incoming hooks nor PR
  'BUILD_ID',
  'DEPLOY_ID',
  'DEPLOY_PRIME_URL',
  'DEPLOY_URL',
  'INCOMING_HOOK_BODY',
  'INCOMING_HOOK_TITLE',
  'INCOMING_HOOK_URL',
  'NETLIFY_BUILD_BASE',
  'NETLIFY_BUILD_LIFECYCLE_TRIAL',
  'NETLIFY_IMAGES_CDN_DOMAIN',
  'PULL_REQUEST',
  'REVIEW_ID',
]
