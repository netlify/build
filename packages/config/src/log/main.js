const { cleanupConfig } = require('./cleanup')
const { logObject, logSubHeader } = require('./logger')
const { cleanupConfigOpts } = require('./options')

// Log options in debug mode.
const logOpts = function(opts, { logs, debug, cachedConfig }) {
  // In production, print those in the first call to `@netlify/config`, not the
  // second one done inside `@netlify/build`
  if (!debug || cachedConfig !== undefined) {
    return
  }

  logSubHeader(logs, 'Initial build environment')
  logObject(logs, cleanupConfigOpts(opts))
}

// Log `defaultConfig` option in debug mode
const logDefaultConfig = function(defaultConfig, { logs, debug, baseRelDir }) {
  if (!debug || defaultConfig === undefined) {
    return
  }

  logSubHeader(logs, 'UI build settings')
  logObject(logs, cleanupConfig({ ...defaultConfig, baseRelDir }))
}

// Log `inlineConfig` option in debug mode
const logInlineConfig = function(initialConfig, { logs, debug }) {
  if (!debug) {
    return
  }

  logSubHeader(logs, 'CLI flags')
  logObject(logs, cleanupConfig(initialConfig))
}

// Log return value of `@netlify/config` in debug mode
const logResult = function({ configPath, buildDir, config, context, branch }, { logs, debug }) {
  if (!debug) {
    return
  }

  logSubHeader(logs, 'Resolved build environment')
  logObject(logs, { configPath, buildDir, context, branch })

  logSubHeader(logs, 'Resolved config')
  logObject(logs, cleanupConfig(config))
}

module.exports = { logOpts, logDefaultConfig, logInlineConfig, logResult }
