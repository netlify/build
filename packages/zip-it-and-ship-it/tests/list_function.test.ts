import { join, normalize, resolve } from 'path'

import { describe, expect, test } from 'vitest'

import { listFunction } from '../src/main.js'

import { FIXTURES_DIR, FIXTURES_ESM_DIR, normalizeFiles } from './helpers/main.js'

describe('listFunction', () => {
  describe('v1', () => {
    test('listFunction does not include runtimeAPIVersion when parseISC false', async () => {
      const mainFile = join(FIXTURES_DIR, 'in-source-config/functions/cron_cjs.js')
      const func = await listFunction(mainFile, {
        parseISC: false,
      })

      expect(func?.runtimeAPIVersion).toBeUndefined()
    })

    test('listFunction includes runtimeAPIVersion when parseISC true', async () => {
      const mainFile = join(FIXTURES_DIR, 'in-source-config/functions/cron_cjs.js')
      const func = await listFunction(mainFile, {
        parseISC: true,
      })

      expect(func?.runtimeAPIVersion).toBe(1)
    })

    test('listFunction includes in-source config declarations', async () => {
      const dir = join(FIXTURES_DIR, 'in-source-config/functions')
      const mainFile = join(dir, 'cron_cjs.js')
      const func = await listFunction(mainFile, {
        parseISC: true,
      })

      expect(func?.extension).toBe('.js')
      expect(func?.inputModuleFormat).toBe('cjs')
      expect(func?.mainFile).toBe(mainFile)
      expect(func?.name).toBe('cron_cjs')
      expect(func?.runtime).toBe('js')
      expect(func?.runtimeAPIVersion).toBe(1)
      expect(func?.schedule).toBe('@daily')
      expect(normalize(func!.srcDir!)).toBe(normalize(dir))
    })

    test('listFunction includes json configured functions with configured properties', async () => {
      const dir = join(FIXTURES_DIR, 'json-config/.netlify/functions-internal')
      const mainFile = join(dir, 'simple.js')
      const func = await listFunction(mainFile, {
        configFileDirectories: [dir],
      })

      expect(func?.displayName).toBe('A Display Name')
      expect(func?.extension).toBe('.js')
      expect(func?.generator).toBe('@netlify/mock-plugin@1.0.0')
      expect(func?.mainFile).toBe(mainFile)
      expect(func?.name).toBe('simple')
      expect(func?.runtime).toBe('js')
      expect(func?.runtimeAPIVersion).toBeUndefined()
      expect(normalize(func!.srcDir!)).toBe(normalize(dir))
    })
  })
  describe('V2 API', () => {
    test('listFunction does not include metadata properties when parseISC false', async () => {
      const mainFile = join(FIXTURES_ESM_DIR, 'v2-api/function.js')
      const func = await listFunction(mainFile, {
        parseISC: false,
      })

      expect(func?.runtimeAPIVersion).toBeUndefined()
    })

    test('listFunction includes metadata properties when parseISC true', async () => {
      const basePath = resolve(FIXTURES_ESM_DIR)
      const mainFile = join(basePath, 'v2-api-esm-ts-aliases/function.ts')
      const func = await listFunction(mainFile, {
        basePath,
        parseISC: true,
      })

      if (!func) {
        throw new Error('Function is not defined')
      }

      const normalizedFunc = normalizeFiles(basePath, func)

      expect(normalizedFunc).toMatchSnapshot()
    })
  })
})
