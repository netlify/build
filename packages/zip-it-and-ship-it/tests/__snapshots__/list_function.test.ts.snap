// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`listFunction > V2 API > listFunction includes metadata properties when parseISC true 1`] = `
{
  "displayName": undefined,
  "excludedRoutes": [],
  "extension": ".ts",
  "generator": undefined,
  "inputModuleFormat": "esm",
  "mainFile": "v2-api-esm-ts-aliases/function.ts",
  "name": "function",
  "routes": [],
  "runtime": "js",
  "runtimeAPIVersion": 2,
  "schedule": undefined,
  "srcDir": "v2-api-esm-ts-aliases",
  "srcFile": undefined,
  "timeout": undefined,
}
`;
